velox/experimental/cudf/exec/CudfHashJoin.cpp

```diff
@@ -468,150 +557,438 @@ void CudfHashJoinProbe::noMoreInput() {
+}
 
-  // TODO pass the input pool !!!
-  // TODO: We should probably subset columns before calling to_cudf_table?
-  // Maybe that isn't a problem if we fuse operators together.
-  auto& rightTable = hashObject_.value().first;
-  auto& hb = hashObject_.value().second;
-  VELOX_CHECK_NOT_NULL(rightTable);
-  if (CudfConfig::getInstance().debugEnabled) {
-    if (rightTable != nullptr)
-      LOG(INFO) << "right_table is not nullptr " << rightTable.get()
-                << " hasValue(" << hashObject_.has_value() << ")" << std::endl;
-    if (hb != nullptr)
-      LOG(INFO) << "hb is not nullptr " << hb.get() << " hasValue("
-                << hashObject_.has_value() << ")" << std::endl;
-  }
+std::unique_ptr<cudf::table> CudfHashJoinProbe::filteredOutput(
+    cudf::table_view leftTableView,
+    cudf::column_view leftIndicesCol,
+    cudf::table_view rightTableView,
+    cudf::column_view rightIndicesCol,
+    std::function<std::vector<std::unique_ptr<cudf::column>>(
+        std::vector<std::unique_ptr<cudf::column>>&&,
+        cudf::mutable_column_view)> func,
+    rmm::cuda_stream_view stream) {
+  auto leftResult =
+      cudf::gather(leftTableView, leftIndicesCol, oobPolicy, stream);
+  auto rightResult =
+      cudf::gather(rightTableView, rightIndicesCol, oobPolicy, stream);
+  auto leftColsSize = leftResult->num_columns();
+  auto rightColsSize = rightResult->num_columns();
 
-  // Special case for null-aware anti join where
-  // build table is not empty, no nulls, and probe table has nulls
-  if (joinNode_->isNullAware() and !joinNode_->filter()) {
-    auto const rightTableHasNulls =
-        cudf::has_nulls(rightTable->view().select(rightKeyIndices_));
-    auto const leftTableHasNulls =
-        cudf::has_nulls(leftTable->view().select(leftKeyIndices_));
-    if (rightTable->num_rows() > 0 and !rightTableHasNulls and
-        leftTableHasNulls) {
-      // drop nulls on probe table
-      leftTable = cudf::drop_nulls(
-          leftTable->view(),
-          leftKeyIndices_,
-          stream,
-          cudf::get_current_device_resource_ref());
-    }
+  std::vector<std::unique_ptr<cudf::column>> joinedCols = leftResult->release();
+  auto rightCols = rightResult->release();
+  joinedCols.insert(
+      joinedCols.end(),
+      std::make_move_iterator(rightCols.begin()),
+      std::make_move_iterator(rightCols.end()));
+
+  auto probeType = joinNode_->sources()[0]->outputType();
+  auto buildType = joinNode_->sources()[1]->outputType();
+  std::vector<velox::RowTypePtr> rowTypes{probeType, buildType};
+  exec::ExprSet exprs({joinNode_->filter()}, operatorCtx_->execCtx());
+  VELOX_CHECK_EQ(exprs.exprs().size(), 1);
+  auto filterEvaluator = ExpressionEvaluator(
+      {exprs.exprs()[0]}, facebook::velox::type::concatRowTypes(rowTypes));
+  auto filterColumns = filterEvaluator.compute(
+      joinedCols, stream, cudf::get_current_device_resource_ref());
+  auto filterColumn = filterColumns[0]->mutable_view();
```

Comment:
This code doesn't compile in main branch, can you fix it please?
Otherwise all current CI is red.

Reason is implicit merge conflict with this PR: https://github.com/facebookincubator/velox/pull/15145 (cudf ExpressionEvaluator was removed)
It was merged first.

@shrshi @devavret